src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/DiagnosticResult.cs

```diff
@@ -224,6 +224,17 @@ public DiagnosticResult WithOptions(DiagnosticOptions options)
 
         public DiagnosticResult WithArguments(params object[] arguments)
         {
+            if (MessageFormat != null)
```

Comment:
💭 What about moving this validation to the `Message` property? Currently the validation will not work for these cases:

1. False exception: `WithArguments` is called before `WithMessageFormat`
2. False non-exception: `WithMessageFormat` is called with 3 placeholders, then `WithArguments` is called with 3 arguments, then `WithMessageFormat` is called again with 2 placeholders.